src/lib/crypto/kimchi_backend/gadgets/curve_params.ml

```diff
@@ -0,0 +1,193 @@
+    ; b : 'field Foreign_field.Element.Standard.t
+    ; gen : 'field Affine.t
+    ; doubles : 'field Affine.t array
+    ; ia : 'field Affine.t ia_points
+    }
+end
+
+let double_bignum_point (curve : t) (point : Affine.bignum_point) :
+    Bignum_bigint.t * Affine.bignum_point =
+  let point_x, point_y = point in
+  (* Compute slope using 1st derivative of sqrt(x^3 + a * x + b)
+     * Note that when a = 0 (e.g. as in the case of secp256k1) we have
+     * one fewer constraint (below).
+  *)
+  let slope =
+    Bignum_bigint.(
+      (* Computes s' = (3 * Px^2  + a )/ 2 * Py *)
+      let numerator =
+        let point_x_squared = pow point_x (of_int 2) % curve.modulus in
+        let point_x3_squared = of_int 3 * point_x_squared % curve.modulus in
+
+        (point_x3_squared + curve.a) % curve.modulus
+      in
+      let denominator = of_int 2 * point_y % curve.modulus in
+
+      (* Compute inverse of denominator *)
+      let denominator_inv =
+        Common.bignum_bigint_inverse denominator curve.modulus
+      in
+      numerator * denominator_inv % curve.modulus)
+  in
+
+  let slope_squared = Bignum_bigint.((pow slope @@ of_int 2) % curve.modulus) in
+
+  (* Compute result's x-coodinate: x = s^2 - 2 * Px *)
+  let result_x =
+    Bignum_bigint.(
+      let point_x2 = of_int 2 * point_x % curve.modulus in
+      (slope_squared - point_x2) % curve.modulus)
+  in
+
+  (* Compute result's y-coodinate: y = s * (Px - x) - Py *)
+  let result_y =
+    Bignum_bigint.(
+      let x_diff = (point_x - result_x) % curve.modulus in
+      let x_diff_s = slope * x_diff % curve.modulus in
+      (x_diff_s - point_y) % curve.modulus)
+  in
+
+  (slope, (result_x, result_y))
+
+let to_circuit_constants (type field)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = field)
+    ?(use_precomputed_gen_doubles = true) (curve : t) : field InCircuit.t =
+  let open Circuit in
+  (* Need to know native field size before we can check if it fits *)
+  Foreign_field.check_modulus_bignum_bigint (module Circuit) curve.modulus ;
+  Foreign_field.check_modulus_bignum_bigint (module Circuit) curve.order ;
+  let order_bit_length = Common.bignum_bigint_bit_length curve.order in
+  let order_minus_one =
+    Bignum_bigint.(if curve.order > zero then curve.order - one else zero)
+  in
+  InCircuit.
+    { bignum = curve
+    ; modulus =
+        Foreign_field.bignum_bigint_to_field_const_standard_limbs
+          (module Circuit)
+          curve.modulus
+    ; order =
+        Foreign_field.bignum_bigint_to_field_const_standard_limbs
+          (module Circuit)
+          curve.order
+    ; order_bit_length
+    ; order_bit_length_const =
+        (let const_len = Field.(constant @@ Constant.of_int order_bit_length) in
+         let var_len =
+           exists Field.typ ~compute:(fun () ->
+               Circuit.Field.Constant.of_int order_bit_length )
+         in
+         Field.Assert.equal const_len var_len ;
```

Comment:
is this `Assert` necessary?

Reply:
Yes.  This is a function that is used to constrain the constants at the start of the circuit so that they do not split up the foreign field addition chains.  To force snarky to constrain it we create a circuit variable and a circuit constant for the same value and then assert equivalence.